### PR TITLE
KAFKA-17543: Enforce that broker.id.generation.enable is not used when migrating to KRaft

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/KafkaServerTest.scala [hunk 1]
```diff
@@ -202,4 +202,16 @@ class KafkaServerTest extends QuorumTestHarness {
     TestUtils.createServer(kafkaConfig)
   }
 
+  @Test
+  def testGeneratedBrokerIdForbiddenDuringMigration(): Unit = {
+    val props = TestUtils.createBrokerConfig(-1, zkConnect)
+    props.put("reserved.broker.max.id", "2000")
```
Comment (mumrah) [inline core/src/test/scala/unit/kafka/server/KafkaServerTest.scala]: Does this test need to set `broker.id.generation.enable`? or is that implicit when `reserved.broker.max.id` is set

**Review by mumrah — COMMENTED** [review]
Thanks @cmccabe! If this is just based on static configs, maybe we should put the validation logic in KafkaConfig. Might be nicer to have all the migration related config validation together.

File: core/src/main/scala/kafka/server/KafkaServer.scala [hunk 2]
```diff
@@ -1135,6 +1132,10 @@ class KafkaServer(
    * @return The brokerId.
    */
   private def getOrGenerateBrokerId(metaPropsEnsemble: MetaPropertiesEnsemble): Int = {
+    if (config.migrationEnabled && config.brokerIdGenerationEnable) {
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaServer.scala]: Should we move this check to https://github.com/apache/kafka/blob/d3936365bf0fc7df86c2632a576081474d923e08/core/src/main/scala/kafka/server/KafkaConfig.scala#L965?

Comment (chia7712) [reply core/src/main/scala/kafka/server/KafkaServer.scala]: BTW, could you please update docs? We need to add `broker.id.generation.enable=false` to the example and revise the following item.

https://github.com/apache/kafka/blob/d3936365bf0fc7df86c2632a576081474d923e08/docs/ops.html#L3969

File: core/src/main/scala/kafka/server/KafkaServer.scala [hunk 3]
```diff
@@ -260,9 +260,6 @@ class KafkaServer(
 
         /* generate brokerId */
         config._brokerId = getOrGenerateBrokerId(initialMetaPropsEnsemble)
-        // Currently, we are migrating from ZooKeeper to KRaft. If broker.id.generation.enable is set to true,
-        // we must ensure that the nodeId synchronizes with the broker.id to prevent the nodeId from being -1,
-        // which would result in a failure during the migration.
         config._nodeId = config.brokerId
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaServer.scala:263]: The `broker.id` must be defined in migration with this patch, so we don't need to update `_nodeId`. Could you please make `nodeId` be immutable again? Or just revert the change of #16540 :smile:

**Review by chia7712 — COMMENTED** [review]
@cmccabe thanks for this patch. This approach was my preference when fixing KAFKA-17077, but I tried to minimize the config changes in migration. Hence, out of curiosity, are there other issues/bugs in migration when `broker.id.generation.enable=true`?

File: core/src/test/scala/unit/kafka/server/KafkaServerTest.scala (hunk 1 above)
Comment (cmccabe) [reply core/src/test/scala/unit/kafka/server/KafkaServerTest.scala]: I removed this, since the config is now enforced in `KafkaConfigTest`.

**Comment by cmccabe** [comment]
> @cmccabe thanks for this patch. This approach was my preference when fixing [KAFKA-17077](https://issues.apache.org/jira/browse/KAFKA-17077), but I tried to minimize the config changes in migration. Hence, out of curiosity, are there other issues/bugs in migration when broker.id.generation.enable=true?

control plane listeners are also not supported in KRaft mode.

File: core/src/main/scala/kafka/server/KafkaServer.scala (hunk 3 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaServer.scala:263]: Yes, fair point. Let's do that later since we're trying to get this specific change into 3.9. In trunk we'll be able to make both immutable.

File: core/src/main/scala/kafka/server/KafkaServer.scala (hunk 2 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaServer.scala]: The existing docs should be OK here, now that I have altered the PR to forbid broker id being set to -1, rather than forbidding `broker.id.generation.enable`.

**Comment by cmccabe** [comment]
I changed the PR slightly to simply forbid setting `broker.id` to -1, rather than forbidding `broker.id.generation.enable = true`. Although the latter approach is more conceptually pure, it would be more disruptive in practice. If someone has a cluster where all their broker IDs are defined, and is not using auto generation, they should not get an error even if they happen to have `broker.id.generation.enable = true`.

(One thing to note is a lot of our tests set `broker.id.generation.enable = true`, although almost none of them actually use it!)

**Comment by chia7712** [comment]
> control plane listeners are also not supported in KRaft mode.

Sorry that I can't catch the point. Could you please share more details to me?

**Comment by cmccabe** [comment]
> > control plane listeners are also not supported in KRaft mode.
>
> Sorry that I can't catch the point. Could you please share more details to me?

The control.plane.listener.name configuration is not supported in KRaft mode

**Review by mumrah — APPROVED** [review]
LGTM

**Review by chia7712 — APPROVED** [review]
LGTM and please revise the title according to last change.

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 4]
```diff
@@ -964,6 +967,8 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
     } else {
       // ZK-based
       if (migrationEnabled) {
+        require(brokerId >= 0,
+          "broker broker.id.generation.enable is incompatible with migration to ZK. Please disable it before enabling migration")
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:971]: @cmccabe is this description "migration to zk" correct?
